services/localfile/server/shred.go

```diff
@@ -0,0 +1,79 @@
+/* Copyright (c) 2019 Snowflake Inc. All rights reserved.
+
+   Licensed under the Apache License, Version 2.0 (the
+   "License"); you may not use this file except in compliance
+   with the License.  You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing,
+   software distributed under the License is distributed on an
+   "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+   KIND, either express or implied.  See the License for the
+   specific language governing permissions and limitations
+   under the License.
+*/
+
+package server
+
+import (
+	"context"
+	pb "github.com/Snowflake-Labs/sansshell/services/localfile"
+	"github.com/Snowflake-Labs/sansshell/services/util"
+	"github.com/Snowflake-Labs/sansshell/telemetry/metrics"
+	"go.opentelemetry.io/otel/attribute"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/types/known/emptypb"
+)
+
+var (
+	localfileShredFailureCounter = metrics.MetricDefinition{
+		Name:        "actions_localfile_shred_failure",
+		Description: "number of failures when performing localfile.Shred",
+	}
+)
+
+func (s *server) Shred(ctx context.Context, req *pb.ShredRequest) (*emptypb.Empty, error) {
+	recorder := metrics.RecorderFromContextOrNoop(ctx)
+
+	if err := util.ValidPath(req.Filename); err != nil {
+		recorder.CounterOrLog(ctx, localfileShredFailureCounter, 1, attribute.String("reason", "invalid_path"))
+		return nil, err
+	}
+
+	shredPath, err := util.Which("shred")
+	if err != nil {
+		recorder.CounterOrLog(ctx, localfileShredFailureCounter, 1, attribute.String("reason", "shred_not_found"))
+		return nil, err
+	}
+	args := make([]string, 0, 4)
+
+	if req.Force {
+		args = append(args, "-f")
+	}
+
+	if req.Zero {
+		args = append(args, "-z")
+	}
+
+	if req.Remove {
+		args = append(args, "-u")
+	}
+
+	args = append(args, req.Filename)
+
+	// we want to fail as soon as possible, if the file is unreadable, opened by someone else or something else happens
+	// to it while we're trying to shred it, we might as well just fail during command execution
+	r, err := util.RunCommand(ctx, shredPath, args)
```

Comment:
To run `shred` command we need to be sure it would be on every VM and every base image where sansshell already in use and/or would be in use. 

What do you think can we avoid using it and make go implementation like it done in https://github.com/pabloandresm/shred/blob/main/shred.go instead?

Reply:
It would be a little bit more complicated to make a proper implementation for HDDs (for SSDs its of course not going to do anything). I don't know much about magnetic storage, but I think there's something to the whole overwrite pattern, judging by the sources here its using pseudo-random (or optionally "true" random hardware based RNG I guess) seeds and patterns for the passes: https://github.com/coreutils/coreutils/blob/master/src/shred.c#L674

Its usually available on every Linux machine, even most lightweight distros include it - both Wolfi and Alpine.
That being said, maybe we can just fail in this case? Alternatively we can just do an RM (it achieves same thing on SSDs).  

As an alternative I could just add a naive basic implementation as a backup for when actual Shred is not available on the system (as unlikely as that is). 

Frankly I'm not sure if it even makes sense to add Shred though, HDDs are of course still used in Clouds, but chances are whatever is being shredded is going to reside on the boot disk (and its would very uncommon for boot disk to be an HDD imo). Maybe the "solution" here is to discard this whole PR?

Reply:
I don't believe we can maintain, long term, a full compatibility contract along the lines of what was suggested above ("on every VM and every base image where sansshell already in use and/or would be in use"). It is not realistic to have such a goal, even golang core libraries have quirks of that nature.

Once I say this, I don't feel it is a good time to go full steam on native shred implementation, I would rather stick to use whatever we have in environment until we know we need a native implementation.